crates/pepper/src/metastore/sqlite.rs

```diff
@@ -0,0 +1,487 @@
+    where
+        F: FnOnce(&dyn MetastoreRow) -> CatalogResult<T> + Send + 'static,
+        T: Send + 'static,
+    {
+        let db_path_owned = self.db_path().to_string();
+        let sql = params.sql.to_string();
+        let param_values = params.params.clone();
+
+        let result = tokio::task::spawn_blocking(move || {
+            let conn = Self::open_connection(&db_path_owned, true)?;
+
+            let params_refs: Vec<Box<dyn rusqlite::ToSql>> =
+                param_values.iter().map(to_sqlite_param).collect();
+
+            let params_slice: Vec<&dyn rusqlite::ToSql> =
+                params_refs.iter().map(|p| p.as_ref()).collect();
+
+            conn.query_row(&sql, params_slice.as_slice(), |row| {
+                let column_count = row.as_ref().column_count();
+                let mut values = Vec::with_capacity(column_count);
+
+                for i in 0..column_count {
+                    let value = row.get_ref(i)?;
+                    values.push(convert_sqlite_value(value));
+                }
+
+                Ok(SqliteRow { values })
+            })
+            .map_err(CatalogError::from)
+            .and_then(|row| f(&row))
+        })
+        .await;
+
+        Self::handle_blocking_result(result, "Query row")
+    }
+
+    async fn query<F, T>(&self, params: QueryParams<'_>, f: F) -> CatalogResult<Vec<T>>
+    where
+        F: Fn(&dyn MetastoreRow) -> CatalogResult<T> + Send + 'static,
+        T: Send + 'static,
+    {
+        let db_path_owned = self.db_path().to_string();
+        let sql = params.sql.to_string();
+        let param_values = params.params.clone();
+
+        let result = tokio::task::spawn_blocking(move || {
+            let conn = Self::open_connection(&db_path_owned, true)?;
+
+            let params_refs: Vec<Box<dyn rusqlite::ToSql>> =
+                param_values.iter().map(to_sqlite_param).collect();
+
+            let params_slice: Vec<&dyn rusqlite::ToSql> =
+                params_refs.iter().map(|p| p.as_ref()).collect();
+
+            let mut stmt = conn.prepare(&sql)?;
+            let rows = stmt.query_map(params_slice.as_slice(), |row| {
+                let column_count = row.as_ref().column_count();
+                let mut values = Vec::with_capacity(column_count);
+
+                for i in 0..column_count {
+                    let value = row.get_ref(i)?;
+                    values.push(convert_sqlite_value(value));
+                }
+
+                Ok(SqliteRow { values })
+            })?;
+
+            let mut results = Vec::new();
+            for row_result in rows {
+                let row = row_result?;
+                results.push(f(&row)?);
+            }
+
+            Ok::<Vec<T>, CatalogError>(results)
+        })
+        .await;
+
+        Self::handle_blocking_result(result, "Query rows")
+    }
+
+    async fn begin_transaction(&self) -> CatalogResult<()> {
```

Comment:
I think this transaction API is bad - its very prone to misuse, we aren't using it at the moment so we should just get rid of it. A better transaction API mimics the existing SQLite/DuckDB pattern: https://docs.rs/rusqlite/latest/rusqlite/struct.Transaction.html